awx_collection/plugins/modules/tower_application.py

```diff
@@ -0,0 +1,159 @@
+      default: "password"
+      choices: ["password", "authorization-code"]
+      type: str
+      required: True
+    client_type:
+      description:
+        - Set to public or confidential depending on how secure the client device is.
+      choices: ["public", "confidential"]
+      type: str
+      required: True
+    organization:
+      description:
+        - Name of organization for application.
+      type: str
+      required: True
+    redirect_uris:
+      description:
+        - Allowed urls list, space separated. Required when authorization-grant-type=authorization-code
+      type: str
+    state:
+      description:
+        - Desired state of the resource.
+      default: "present"
+      choices: ["present", "absent"]
+      type: str
+    skip_authorization:
+      description:
+        - Set True to skip authorization step for completely trusted applications.
+      default: false
+      type: bool
+'''
+
+
+EXAMPLES = '''
+- name: Add Foo application
+  tower_application:
+    name: "Foo"
+    description: "Foo bar application"
+    organization: "test"
+    state: present
+    authorization-grant-type: password
+    client-type: public
+
+- name: Add Foo application
+  tower_application:
+    name: "Foo"
+    description: "Foo bar application"
+    organization: "test"
+    state: present
+    authorization-grant-type: authorization-code
+    client-type: confidential
+    redirect_uris: http://tower.com/api/v2/
+'''
+
+import time
+
+from ..module_utils.tower_api import TowerModule
+
+
+def main():
+    # Any additional arguments that are not fields of the item can be added here
+    argument_spec = dict(
+        name=dict(required=True),
+        description=dict(),
+        authorization_grant_type=dict(required=True),
+        client_type=dict(required=True, choices=['public', 'confidential']),
+        organization=dict(required=True),
+        redirect_uris=dict(type="list", elements='str'),
+        state=dict(choices=['present', 'absent'], default='present'),
+        skip_authorization=dict(type=bool)
+    )
+
+    # Create a module for ourselves
+    module = TowerModule(argument_spec=argument_spec)
+
+    # Extract our parameters
+    name = module.params.get('name')
+    description = module.params.get('description')
+    authorization_grant_type = module.params.get('authorization_grant_type')
+    client_type = module.params.get('client_type')
+    organization = module.params.get('organization')
+    redirect_uris = ' '.join(module.params.get('redirect_uris'))
+    state = module.params.get('state')
+
+    # Attempt to look up the related items the user specified (these will fail the module if not found)
+    org_id = module.resolve_name_to_id('organizations', organization)
+
+    # Attempt to look up application based on the provided name and org ID
+    application = module.get_one('applications', **{
+        'data': {
+            'name': name,
+            'organization': org_id
+        }
+    })
```

Comment:
So we have finally addressed this. The get_one method now takes an optional argument called name_or_id. If you change this code snippet to:
```
    application = module.get_one('applications', name_or_id=name, **{
        'data': {
            'organization': org_id
        }
    })
```

That will allow for either a name or an ID to be passed in.